app/com/arpnetworking/metrics/portal/scheduling/JobRepository.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2018 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.portal.scheduling;
+
+import models.internal.scheduling.Job;
+
+import java.time.Instant;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * A storage medium for {@link Job}s. Essentially a Map that mints unique keys for new Job values.
+ *
+ * @param <T> The type of result produced by the {@link Job}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public interface JobRepository<T> {
```

Comment:
This needs a `jobCompleted(Job<T> job, Job.State state, T result)`. I had considered renaming this since completed doesn't really make sense when the status update is `RUNNING` but we can cross that bridge when we get there with a rename. 

I would also consider if `result` should be `@Nullable` since it seems nonsensical to expect a result when the Job failed.